tools/generator/Utilities/Report.cs

```diff
@@ -1,97 +1,113 @@
-		public const int WarningCtor = 0x8300;
-		public const int WarningField = 0x8400;
-		public const int WarningFieldNameCollision = 0x8401;
-		public const int WarningDuplicateField = 0x8402;
-		public const int WarningInterfaceGen = 0x8500;
-		public const int WarningParser = 0x8600;
-		public const int WarningReturnValue = 0x8700;
-		public const int WarningParameter = 0x8800;
-		public const int WarningGenBaseSupport = 0x8900;
-		public const int WarningApiFixup = 0x8A00;
-		public const int WarningGenericParameterDefinition = 0x8B00;
-		public const int WarningGenBase = 0x8C00;
-		public const int WarningMethodBase = 0x8D00;
-		public const int WarningAnnotationsProvider = 0x8E00;
-
 		public static int? Verbosity { get; set; }
 
-		public static void Error (int errorCode, string format, params object[] args)
+		public class LocalizedMessage
 		{
-			Error (errorCode, null, null, -1, -1, format, args);
-		}
+			public int Code { get; set; }
+			public string Value { get; set; }
 
-		public static void Error (int errorCode, string sourceFile, int line, int column, string format, params object [] args)
-		{
-			Error (errorCode, null, sourceFile, line, column, format, args);
+			public LocalizedMessage (int code, string value)
+			{
+				Code = code;
+				Value = value;
+			}
 		}
 
-		public static void Error (int errorCode, Exception innerException, string format, params object [] args)
-		{
-			Error (errorCode, innerException, null, -1, -1, format, args);
-		}
+		public static LocalizedMessage ErrorFailedToRemoveConstants => new LocalizedMessage (4000, Java.Interop.Localization.Resources.Generator_BG4000);
+		public static LocalizedMessage ErrorFailedToProcessEnumMap => new LocalizedMessage (4100, Java.Interop.Localization.Resources.Generator_BG4100);
+		public static LocalizedMessage ErrorFailedToProcessMetadata => new LocalizedMessage (4200, Java.Interop.Localization.Resources.Generator_BG4200);
+		public static LocalizedMessage ErrorRemoveNodeInvalidXPath => new LocalizedMessage (4301, Java.Interop.Localization.Resources.Generator_BG4300);
+		public static LocalizedMessage ErrorAddNodeInvalidXPath => new LocalizedMessage (4302, Java.Interop.Localization.Resources.Generator_BG4300);
+		public static LocalizedMessage ErrorChangeNodeInvalidXPath => new LocalizedMessage (4303, Java.Interop.Localization.Resources.Generator_BG4300);
+		public static LocalizedMessage ErrorAttrInvalidXPath => new LocalizedMessage (4304, Java.Interop.Localization.Resources.Generator_BG4300);
+		public static LocalizedMessage ErrorMoveNodeInvalidXPath => new LocalizedMessage (4305, Java.Interop.Localization.Resources.Generator_BG4300);
+		public static LocalizedMessage ErrorRemoveAttrInvalidXPath => new LocalizedMessage (4306, Java.Interop.Localization.Resources.Generator_BG4300);
+		public static LocalizedMessage ErrorMissingAttrName => new LocalizedMessage (4307, Java.Interop.Localization.Resources.Generator_BG4307);
```

Comment:
I don't have much of an opinion here.  I felt this approach mimicked the existing approach in `generator`.  I think it's more about finding where an error/warning message *code* comes from rather than the *text*.

For example, say someone reported a `BG4306` error and you want to find where in the code it comes from.  There is no `Resources.Generator_BG4306` because we don't want to duplicate strings in `resources.resx` because that causes extra translator work.

```
public static LocalizedMessage ErrorMoveNodeInvalidXPath => new LocalizedMessage (4305, Java.Interop.Localization.Resources.Generator_BG4300);
public static LocalizedMessage ErrorRemoveAttrInvalidXPath => new LocalizedMessage (4306, Java.Interop.Localization.Resources.Generator_BG4300);
```

Searching for `BG4306` would technically work because of the comment, but most `LogCodedError ()` invocations don't have a comment.  So for other cases, you could at least you can scroll through `Report.cs` and quickly find the error message.

So it feels like the benefit of this is:
- It ties message codes to message strings, since they aren't a 1:1 mapping.
- Provides a single place where you can see all codes that can be logged.

But still, I'm not very attached to this approach.  Any approach feels like a pretty minor detail that will rarely be interacted with. 🤷

Reply:
Maybe the "friendly name" is an annoying additional thing to name, and we should just call them the message code to make it easier to find:

```
public static LocalizedMessage BG4305 => new LocalizedMessage (4305, Java.Interop.Localization.Resources.Generator_BG4300);

instead of

public static LocalizedMessage ErrorMoveNodeInvalidXPath => new LocalizedMessage (4305, Java.Interop.Localization.Resources.Generator_BG4300);
```

The friendly naming was another case of "following the already established approach".

Reply:
Ooh, yeah, that approach you mentioned for cases where the same string is used for two error codes is a nice touch. In the one case where an identical string was used by two different codes in xamarin-android, I just did the naive thing and made two copies of the string under separate resource names. I believe the translation team has a process where they check new strings against a dictionary of existing translations so that they avoid duplicating translation work, but all the nicer to take an approach in the _.resx_ file that avoids that question entirely.

On naming, the current descriptive names work fine for me, but I'd also be a-OK with having the error codes be the names, so whichever one you like best will be good.